File: samplecpp/src/conn/GeneralChannel.cpp
#include "GeneralChannel.h"
#include "../Const.h"

void GeneralChannel::initConfig(Json config) {
    this->config = config;

    client = new SocketClient(
            config[Const::main_config][Const::master_ip],
            (int) config[Const::main_config][Const::master_port],
            NonBlockingSocket);


    while (true) {
        std::string line = client->ReceiveLine();
        if (line.size()) {
            Print("id recved: %s", line.data());
            Json exConfig = Json::parse(line);
            this->config[Const::index] = exConfig[Const::index];
            this->config[Const::max_channel_conn] = exConfig[Const::max_channel_conn];
            break;
        } else {
            Sleep(50);
        }
    }
    Print("init finish.")
}

void GeneralChannel::send(Json message, int target) {
    message[Const::target_id] = target;
    Print("msg send: %s", message.dump().data());
    client->SendLine(message.dump());
}

std::vector<Json> GeneralChannel::recv() {
    std::vector<Json> result;
    while (true) {
        std::string line = client->ReceiveLine();
        if (line.size()) {
            Print("msg recv: %s", line.data());
            Json message = Json::parse(line);
            message[Const::recv_time] = Const::curTime();
            result.push_back(message);
        } else {
            break;
        }
    }
    return result;
}

int GeneralChannel::getId() {
    return config[Const::index];
}

Json GeneralChannel::getConfig(std::vector<std::string> &keys){
    Json ans = config;
    for (std::string key: keys) {
        ans = ans[key];
    }
    return std::move(ans);
}

File: pysample/src/conn/GeneralChannel.py
import json
import time
import Channel
import SocketClient
from src import Const


class GeneralChannel(Channel.Channel):

    def __init__(self):
        self.client = None
        self.config = None

    def _init_main_config(self):
        self.client = SocketClient.SocketClient(self.config[Const.main_config][Const.master_ip],
                                                self.config[Const.main_config][Const.master_port])
        while True:
            line = self.client.read_line()
            if len(line):
                print "id recved : %s" % (line,)
                ex_config = json.loads(line)
                self.config[Const.index] = ex_config[Const.index]
                self.config[Const.max_channel_conn] = ex_config[Const.max_channel_conn]
                break
            else:
                time.sleep(0.05)

    def init_config(self, config):
        self.config = config
        self._init_main_config()

    def send(self, message, target_id):
        message[Const.target_id] = target_id
        print "msg send : %s" % (message, )
        self.client.println(json.dumps(message))

    def recv(self):
        result = []
        while True:
            line = self.client.read_line()
            if len(line):
                print "msg recv: %s" % (line, )
                message = json.loads(line)
                message[Const.recv_time] = time.time()
                result.append(message)
            else:
                break
        return result

    def get_id(self):
        return self.config[Const.index]

    def get_config(self, args):
        obj = self.config
        for arg in args:
            obj = obj[arg]
        return obj
